javalib/src/main/scala/java/util/LinkedList.scala

```diff
@@ -0,0 +1,373 @@
+    else peekFirst()
+
+  def getLast(): E =
+    if (_size <= 0) throw new NoSuchElementException()
+    else peekLast()
+
+  def removeFirst(): E = {
+    if (isEmpty()) throw new NoSuchElementException()
+    val ret = head.value
+
+    val oldHead = head
+    head = oldHead.next
+
+    if (head ne null)
+      head.prev = null
+    else
+      head = last
+
+    _size -= 1
+    ret
+  }
+
+  def removeLast(): E = {
+    if (isEmpty()) throw new NoSuchElementException()
+    val ret = last.value
+
+    val oldLast = last
+    last = oldLast.prev
+
+    if (last ne null)
+      head.next = null
+    else
+      last = head
+
+    _size -= 1
+    ret
+  }
+
+  def addFirst(e: E): Unit = {
+    val oldHead = head
+
+    head = new Node(e, next = oldHead)
+
+    _size += 1
+
+    if (oldHead ne null)
+      oldHead.prev = head
+    else
+      last = head
+  }
+
+  def addLast(e: E): Unit = {
+    val oldLast = last
+
+    last = new Node(e, prev = oldLast)
+
+    _size += 1
+
+    if (oldLast ne null)
+      oldLast.next = last
+    else
+      head = last
+  }
+
+  override def contains(o: Any): Boolean =
+    iterator().exists(_ === o)
+
+  override def size(): Int =
+    _size
+
+  override def add(e: E): Boolean = {
+    addLast(e)
+    true
+  }
+
+  override def remove(o: Any): Boolean =
+    _removeOccurrence(listIterator, o)
+
+  override def addAll(c: Collection[_ <: E]): Boolean = {
+    val iter = c.iterator
+    val changed = iter.hasNext()
+    while (iter.hasNext())
+      addLast(iter.next())
+
+    changed
+  }
+
+  override def clear(): Unit = {
+    head = null
+    last = null
+    _size = 0
+  }
+
+  private def getNodeAt(index: Int): Node[E] = {
+    if (index == 0) head
+    else if (index == size - 1) last
+    else {
+      var current: Node[E] = null
+      if (index <= size/2) {
+        current = head
+        for (_ <- 0 until index) current = current.next
+      } else {
+        current = last
+        for (_ <- index until (size - 1)) current = current.prev
+      }
+      current
+    }
+  }
+
+  override def get(index: Int): E = {
+    checkIndexInBounds(index)
+    getNodeAt(index).value
+  }
+
+  override def set(index: Int, element: E): E = {
+    checkIndexInBounds(index)
+    val node = getNodeAt(index)
+    val oldValue = node.value
+    node.value = element
+    oldValue
+  }
+
+  private def addNode(nextNode: Node[E], e: E): Unit = {
+    if (nextNode eq head) addFirst(e)
+    else if (nextNode eq null) addLast(e)
+    else {
+
+      val node = new Node(e, prev = nextNode.prev, next = nextNode)
+      nextNode.prev.next = node
+      nextNode.prev = node
+
+      _size += 1
+    }
+  }
+
+  override def add(index: Int, element: E): Unit = {
+    checkIndexOnBounds(index)
+    addNode(getNodeAt(index), element)
+  }
+
+  private def removeNode(node: Node[E]): E = {
+    if (node eq head) removeFirst()
+    else if (node eq last) removeLast()
+    else {
+
+      node.prev.next = node.next
```

Comment:
Spurious blank line. (In general, there should never be a blank line after the `{` or before the `}` of an if/while/for/lambda. TODO for the coding style guide, I guess.)